# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

In [10]:
#whiteboard
#You must implement a function that return the difference between the biggest and the smallest value in a list(lst) received as parameter.
#The list(lst) contains integers, that means it may contain some negative numbers.
#If the list is empty or contains a single element, return 0.
#The list(lst) is not sorted.
#max_diff([1, 2, 3, 4]) # return 3, because 4 - 1 == 3
#max_diff([1, 2, 3, -4]) # return 7, because 3 - (-4) == 7

def difference(lst):
    if lst == []:
        return 0
    
    a= max(lst)
    b= min(lst)
    c = (a-b)
    return c 

difference([])
    

0

#### Syntax

In [11]:
#user defined function
def add_two(num):
    return num + 2

add_two(4)

6

In [14]:
#same function as a Lambbda function
(lambda num: num + 2)(5)

7

#### Saving to a Variable

In [15]:
f_test = lambda num: num + 2

f_test(5)

7

#### Multiple Inputs

In [18]:
#Multiples with no variable

(lambda x, y, z: x*y*z)(4,2,3)

24

In [19]:
#multiple inputs with a variable attached
mult_func = lambda x, y, z: x*y*z

mult_func(3, 4, 6)

72

#### Passing a Lambda into a Function

In [28]:
def apply_func(f, num):
    """
    f expects a function
    num expects a number
    apply_func will run function f and pass in num as an argument
    """
    
    return f, num

#def my_func(x):
 #   return x * x

apply_func(lambda x: x*x, 4) #lamda 4: 4 *4 =16

(<function __main__.<lambda>(x)>, 4)

#### Returning a Lambda from a Function

In [32]:
def multiply_test(num):
    return num * 4

type(multiply_test(3))

#funtion that returns function

def return_func():
    def multiply(num):
        return num * 2
    return multiply

a = return_func()
a(5)

print(return_func)
print(return_func())

#Lambda function returned from a regular function
def return_lambda(b, c):
    return lambda x:x * (b + c)

r_lamb = return_lambda(2, 4)
print(r_lamb)
r_lamb(5)

<function return_func at 0x00000201715848B0>
<function return_func.<locals>.multiply at 0x0000020171584820>
<function return_lambda.<locals>.<lambda> at 0x0000020171584CA0>


30

#### If Statements within Lambdas

In [33]:
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(10))

print(f_condition(12))

10
12
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [36]:
new_test = lambda num: num *2 if num > 20 else num * 3 if num < 5 else num ** 2
print(new_test(22))
print(new_test(2))
print(new_test(10))

44
6
100


In [37]:
f = lambda num: num ** 3

print(f(2))

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [43]:
#map(function to apply, iterable funtion applies to( a list or tuple or set or dictionary))

#with a predefined function
def square(num):
    return num ** 2

numbers = [4, 22, 44, 2, 4, 8, 3, 66]

squared_nums = list(map(square, numbers))
squared_nums

[16, 484, 1936, 4, 16, 64, 9, 4356]

In [44]:
def square(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num, num2

numbers = [4, 22, 44, 2, 4, 8, 3, 66]
more_numbers = [2, 10, 22, 7]


squared_nums = list(map(square, numbers, more_numbers))
squared_nums

[(16, 4), (22, 10), (44, 22), (4, 49)]

#### Using Lambda's with Map

In [50]:
#map(lambda func, iterable)

square_lamb_nums = map(lambda num, num2: (num**2, num2**2) if num < 10 and num2 < 10 else (num, num2), numbers, more_numbers)
list(squared_lamb_nums)

NameError: name 'squared_lamb_nums' is not defined

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [60]:
#my solution
my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]
#extwo = map(lambda x: (x * 2 - 1,my_nums) 
#          
#print(my_nums)

#brians solution
list(map(lambda num:num*2 - 1, my_nums))

[9, 15, 19, 29, 47, 83, 33, 27, 5, 17]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [67]:
#we'll return a true or false on values to filter things out of a list

def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
num_list = [1, 2, 3, 4, 7, 88, 44, 33]
list(filter(is_even, num_list))



[2, 4, 88, 44]

In [68]:
def names_that_start_with_a(name):
    if name[0].lower() == 'a':
        return True
    return False


names = ["bob", "amber", "andrew", "frank"]
list(filter(names_that_start_with_a, names))

['amber', 'andrew']

In [69]:
def truthyness(num):
    if num > 10:
        return "Hello World"
    else:
        return ""
    
list(filter(truthyness, [1,2,3,15,16]))

[15, 16]

#### Using Lambda's with Filter()

In [71]:
name = "arian"
name[0].lower() == "a"

True

In [72]:
new_names_lamb = list(filter(lambda name: name[0].lower() =="a", names))
new_names_lamb

['amber', 'andrew']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [82]:
#import statistics
#my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]
#def stats (my_nums):
#    x= statistics.mean(my_nums)
#    y = list(filter(my_nums>= x(my_nums)))
#    return y
#
#stats([5, 8, 10, 15, 24, 42, 17, 14, 3, 9])

In [87]:
import statistics
my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]

mean = statistic.mean(my_nums)

def is_greater_than_mean(num):
    if num > mean:
        return True
    else:
        return False

list(filter(is_greater_than_mean, my_nums))

NameError: name 'statistic' is not defined

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [3]:
from functools import reduce

#(reduce function iterables)
def add_nums(accumulated_value, updates_value):
    return accumulated_value + updates_value

list_1 =[2, 4, 6, 8, 10]

total = 0
for num in list_1:
    total = add_nums(total, num)
    
print(total)



30


In [8]:
reduce(add_nums, list_1)
print(sum(list_1))

30


In [10]:
def sub_nums(accumulated_value, updates_value):
    return accumulated_value - updates_value
list_1 =[2, 4, 6, 8, 10]

total = 0
for num in list_1:
    total = sub_nums(total, num)
    
print(total)


-30


#### Using Lambda's with Reduce()

In [12]:
my_list = [1, 2, 3, 4, 5]

reduce(lambda x,y: x * y , my_list)


120

In [13]:
total = 1
for num in my_list:
    total *= num
    
total

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [14]:
my_list = [1, 2, 3, 4, 5]

reduce(lambda x,y: x * y , my_list)

120

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [17]:
def add_nums(num):
    # Set base case for recursive function
    if num <=1 :
        print("add_nums(1)=1")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1)=1


15

#### Writing a Factorial Function

In [22]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num - 1)
    
factorial(5)#can take a very long time depending on size of number

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [25]:
def sub_nums(num):
    if num <=1:
        print(f'sub_nums(1) = 1')
        return num
    else:
        print(f'sub_nums({num}) = {num} - sub_nums({num-1})')
        return num - sub_nums(num - 1)
sub_nums(10)

sub_nums(10) = 10 - sub_nums(9)
sub_nums(9) = 9 - sub_nums(8)
sub_nums(8) = 8 - sub_nums(7)
sub_nums(7) = 7 - sub_nums(6)
sub_nums(6) = 6 - sub_nums(5)
sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums(1) = 1


5

In [30]:
class Example:
    def __init__(self, value):
        self.value = value
        self.next = None
a = Example(1)
b = Example(2)
c = Example(3)
a.next = b
b.next = c
def print_last_item_of_linked_list(head_of_linked_list):
    if not head_of_linked_list.next:
        print(head_of_linked_list.value)
    else:
        print_last_item_of_linked_list(head_of_linked_list.next)
print_last_item_of_linked_list(a)

3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [31]:
def my_range(stop, start=0, step=1):
    while start < stop:
        yield start
        start += step
        
test = my_range(10)

In [32]:
list(test) #once you print a generator with "yeild" it then is used up and con only be printed once

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [52]:
def nums_squared(num):
    while num> 0:
        yield num**2
        num -= 1
        
for s in nums_squared(10):
    print(s)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [56]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def myFunc(x):
    txt = x
    txtsti = txt.strip()
    spaces = ("")
    if txtsti == spaces:
        return False
    else:
        return True

names = filter(myFunc, places)

for x in names:
  print(x)

#did this one all on my own!


Argentina
San Diego
Boston
New York


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [99]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"] #-1 takes the last index


author.sort(key=lambda x: x.split(" ")[-1].title(),reverse=False)#tatyana helped me with this. I don't understand the reverse = false

print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [128]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

ftoc = list(map(lambda x:(x[0],((9/5)*x[1]*(9/5)+32))
       
print(places)


SyntaxError: invalid syntax (<ipython-input-128-bca7ace17eb9>, line 6)

# Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [44]:
# ¯\_(ツ)_/¯ 
def sub_nums(num):
    if num >=200:
        return num + num
    else:
        print(f'Iteration {num}: {num+1}')
        return (num - 1) + sub_nums(num - 2)
sub_nums(2)

Iteration 2: 3
Iteration 0: 1
Iteration -2: -1
Iteration -4: -3
Iteration -6: -5
Iteration -8: -7
Iteration -10: -9
Iteration -12: -11
Iteration -14: -13
Iteration -16: -15
Iteration -18: -17
Iteration -20: -19
Iteration -22: -21
Iteration -24: -23
Iteration -26: -25
Iteration -28: -27
Iteration -30: -29
Iteration -32: -31
Iteration -34: -33
Iteration -36: -35
Iteration -38: -37
Iteration -40: -39
Iteration -42: -41
Iteration -44: -43
Iteration -46: -45
Iteration -48: -47
Iteration -50: -49
Iteration -52: -51
Iteration -54: -53
Iteration -56: -55
Iteration -58: -57
Iteration -60: -59
Iteration -62: -61
Iteration -64: -63
Iteration -66: -65
Iteration -68: -67
Iteration -70: -69
Iteration -72: -71
Iteration -74: -73
Iteration -76: -75
Iteration -78: -77
Iteration -80: -79
Iteration -82: -81
Iteration -84: -83
Iteration -86: -85
Iteration -88: -87
Iteration -90: -89
Iteration -92: -91
Iteration -94: -93
Iteration -96: -95
Iteration -98: -97
Iteration -100: -99
Iteration -102: -101
Iterati

Iteration -2688: -2687
Iteration -2690: -2689
Iteration -2692: -2691
Iteration -2694: -2693
Iteration -2696: -2695
Iteration -2698: -2697
Iteration -2700: -2699
Iteration -2702: -2701
Iteration -2704: -2703
Iteration -2706: -2705
Iteration -2708: -2707
Iteration -2710: -2709
Iteration -2712: -2711
Iteration -2714: -2713
Iteration -2716: -2715
Iteration -2718: -2717
Iteration -2720: -2719
Iteration -2722: -2721
Iteration -2724: -2723
Iteration -2726: -2725
Iteration -2728: -2727
Iteration -2730: -2729
Iteration -2732: -2731
Iteration -2734: -2733
Iteration -2736: -2735
Iteration -2738: -2737
Iteration -2740: -2739
Iteration -2742: -2741
Iteration -2744: -2743
Iteration -2746: -2745
Iteration -2748: -2747
Iteration -2750: -2749
Iteration -2752: -2751
Iteration -2754: -2753
Iteration -2756: -2755
Iteration -2758: -2757
Iteration -2760: -2759
Iteration -2762: -2761
Iteration -2764: -2763
Iteration -2766: -2765
Iteration -2768: -2767
Iteration -2770: -2769
Iteration -2772: -2771
Iteration -

Iteration -4180: -4179
Iteration -4182: -4181
Iteration -4184: -4183
Iteration -4186: -4185
Iteration -4188: -4187
Iteration -4190: -4189
Iteration -4192: -4191
Iteration -4194: -4193
Iteration -4196: -4195
Iteration -4198: -4197
Iteration -4200: -4199
Iteration -4202: -4201
Iteration -4204: -4203
Iteration -4206: -4205
Iteration -4208: -4207
Iteration -4210: -4209
Iteration -4212: -4211
Iteration -4214: -4213
Iteration -4216: -4215
Iteration -4218: -4217
Iteration -4220: -4219
Iteration -4222: -4221
Iteration -4224: -4223
Iteration -4226: -4225
Iteration -4228: -4227
Iteration -4230: -4229
Iteration -4232: -4231
Iteration -4234: -4233
Iteration -4236: -4235
Iteration -4238: -4237
Iteration -4240: -4239
Iteration -4242: -4241
Iteration -4244: -4243
Iteration -4246: -4245
Iteration -4248: -4247
Iteration -4250: -4249
Iteration -4252: -4251
Iteration -4254: -4253
Iteration -4256: -4255
Iteration -4258: -4257
Iteration -4260: -4259
Iteration -4262: -4261
Iteration -4264: -4263
Iteration -

Iteration -5642: -5641
Iteration -5644: -5643
Iteration -5646: -5645
Iteration -5648: -5647
Iteration -5650: -5649
Iteration -5652: -5651
Iteration -5654: -5653
Iteration -5656: -5655
Iteration -5658: -5657
Iteration -5660: -5659
Iteration -5662: -5661
Iteration -5664: -5663
Iteration -5666: -5665
Iteration -5668: -5667
Iteration -5670: -5669
Iteration -5672: -5671
Iteration -5674: -5673
Iteration -5676: -5675
Iteration -5678: -5677
Iteration -5680: -5679
Iteration -5682: -5681
Iteration -5684: -5683
Iteration -5686: -5685
Iteration -5688: -5687
Iteration -5690: -5689
Iteration -5692: -5691
Iteration -5694: -5693
Iteration -5696: -5695
Iteration -5698: -5697
Iteration -5700: -5699
Iteration -5702: -5701
Iteration -5704: -5703
Iteration -5706: -5705
Iteration -5708: -5707
Iteration -5710: -5709
Iteration -5712: -5711
Iteration -5714: -5713
Iteration -5716: -5715
Iteration -5718: -5717
Iteration -5720: -5719
Iteration -5722: -5721
Iteration -5724: -5723
Iteration -5726: -5725
Iteration -

RecursionError: maximum recursion depth exceeded while calling a Python object

5
